source/dem/localize_contacts.cc

```diff
@@ -27,216 +28,196 @@ localize_contacts(
+                std::find(object_contact_candidates->begin(),
+                          object_contact_candidates->end(),
+                          particle_id);
 
-          if (search_iterator !=
-              particle_floating_wall_contact_candidate_element->end())
-            {
-              particle_floating_wall_contact_candidate_element->erase(
-                search_iterator);
-              ++particle_floating_wall_map_iterator;
+              // Particle-particle pairs are removed from the list of the new
+              // board search candidates or from the list of the history of the
+              // fine search candidate list depending on the last checks
+              if (search_iterator != contact_candidate_element->end())
+                {
+                  // The current 2nd is a candidate to the current particle.
+                  // The 2nd particle is then removed from the new board search
+                  // particle-particle contact candidates list but kept as
+                  // history in the fine search list
+                  contact_candidate_element->erase(search_iterator);
+                  ++adjacent_map_iterator;
+                }
+              else if (search_object_to_particle_iterator !=
+                         object_contact_candidates->end() &&
+                       contact_type == ContactType::local_particle_particle)
+                {
+                  // The current particle from history list is still a candidate
+                  // to the 2nd particle and the contact is local
+                  // The 2nd particle is then removed from the new board search
+                  // particle-particle contact candidates list but kept as
+                  // history in the fine search list
+                  object_contact_candidates->erase(
+                    search_object_to_particle_iterator);
+                  ++adjacent_map_iterator;
+                }
+              else if (search_object_to_particle_iterator !=
+                         object_contact_candidates->end() &&
+                       contact_type == ContactType::ghost_particle_particle)
+                {
+                  // The current particle from history list is still a candidate
+                  // to the 2nd particle and the contact is not local
+                  // The 2nd particle is then removed from the history in the
+                  // fine search list particle-particle contact candidates list
+                  // but kept in new board search list
+                  adjacent_pairs_content->erase(adjacent_map_iterator++);
+                }
+              else
+                {
+                  // Current particles are no longer contact candidates.
+                  // The particle is then removed as a candidate for fine search
+                  // contact candidates list
+                  adjacent_pairs_content->erase(adjacent_map_iterator++);
+                }
             }
           else
```

Comment:
instead of an else, just another if constexpr. This way it is more explicit which case this represents.